rfd/00XY-session-locking.md

```diff
@@ -0,0 +1,250 @@
+the default period of 10 minutes defined in `defaults.LowResPollingPeriod`)
+a `types.Watcher`-based algorithm should be preferred.
+
+To be able to distinguish session locks received from a remote (root) cluster,
+the root cluster should send the `SessionLock` resource a resource label of the
+form:
+```
+teleport.dev/replicated-from: <root-cluster-name>
+```
+
+### Scenarios
+
+#### Creating a permanent lock
+
+```
+$ tctl sessions lock --user=foo@example.com --message="Suspicious activity."
+Created a session lock with ID "dc7cee9d-fe5e-4534-a90d-db770f0234a1".
+```
+
+This locks out `foo@example.com` without automatic expiration.
+The lock can be lifted by `tctl rm lock/dc7cee9d-fe5e-4534-a90d-db770f0234a1`.
+
+The above locking command would be equivalent to
+```sh
+tctl create <<EOF
+kind: session_lock
+metadata:
+  name: dc7cee9d-fe5e-4534-a90d-db770f0234a1
+spec:
+  message: "Suspicious activity."
+  target:
+    user: foo@example.com
+version: v2
+EOF
+```
+
+The showed YAML would also correspond to the output of `tctl get lock/dc7cee9d-fe5e-4534-a90d-db770f0234a1`.
+
+#### Creating a lock with expiry
+
+```
+$ tctl sessions lock --role=developers --message="Cluster maintenance." --expires-in=10h
+Created a session lock with ID "dc7cee9d-fe5e-4534-a90d-db770f0234a1".
+```
+
+This locks out users with the role `developers` for the next 10 hours.
+
+Assuming the time at which the command is issued on 2021-06-14 at 12:27 UTC,
+the above locking command would be equivalent to both
+```sh
+tctl create <<EOF
+kind: session_lock
+metadata:
+  name: dc7cee9d-fe5e-4534-a90d-db770f0234a1
+  expires: "2021-06-14T22:27:00Z"   # RFC3339
+spec:
+  target:
+    role: developers
+  message: "Cluster maintenance."
+version: v2
+EOF
+```
+and
+```sh
+tctl sessions lock --role=developers --message="Cluster maintenance." --expires="2021-06-14T22:27:00Z"
+```
+
+#### Creating a delayed lock
+
+```
+$ tctl sessions lock --cluster=leaf --effective-from="2021-06-14T22:27:00Z"
+Created a session lock with ID "dc7cee9d-fe5e-4534-a90d-db770f0234a1".
+```
+
+This locks out users accessing the cluster `leaf` effective from 2021-06-14 22:27 UTC.
+
+The above locking command would be equivalent to
+```sh
+tctl create <<EOF
+kind: session_lock
+metadata:
+  name: dc7cee9d-fe5e-4534-a90d-db770f0234a1
+spec:
+  target:
+    cluster: leaf
+  effective_from: "2021-06-14T22:27:00Z"   # RFC3339
+version: v2
+EOF
+```
+and, assuming the time at which the command is issued on 2021-06-14
+at 12:27 UTC, to
+```sh
+tctl sessions lock --cluster=leaf --effective-in=10h
+```
+
+#### Generation of new user certificates prevented
+
+Attempts to generate a new user certificate (`tsh login`, `tctl auth sign`)
+while a session lock targeting role `developers` is in force will result in the
+following error:
+
+```
+ERROR: session lock targeting role "developers" is in force
+```
+
+#### Live SSH session terminated
```

Comment:
This would also impact all other types of sessions, k8s, application and database sessions, right?

Reply:
I'm not sure there is such a thing as "live app session". This particular example was meant to give an analogy to how disconnecting expired certs works, and the latter is not implemented for app access.

Reply:
We should reject all requests then - I think `session` part is misleading for us, because the lock should be locking every single interaction for blocked identity. May be call it `IdentityLock` @fspmarshall does it resonate?

Reply:
I would still probably call it just `Lock` to be honest - pretty clear:

`lock: best_effort | strict ` 
`tctl locks ls`
`tctl lock --mfa-device=id`

Reply:
Yeah, just `Lock` seems good to me too.  Simple.  Since users need to specify what they are locking anyhow, supporting new kinds of locks shouldn't require breaking any interactions.